src/herder/TransactionQueue.cpp

```diff
@@ -43,28 +44,123 @@ TransactionQueue::tryAdd(TransactionFrameBasePtr tx)
+    AccountTransactions emptyAccountTransactions;
+    auto pendingIter = mPendingTransactions.find(tx->getSourceID());
+    auto const& pending = pendingIter != mPendingTransactions.end()
+                              ? pendingIter->second
+                              : emptyAccountTransactions;
+    auto const& transactions = pending.mTransactions;
+
+    auto oldTxIter = transactions.end();
+    if (!transactions.empty())
+    {
+        int64_t firstSeq = transactions.front()->getSeqNum();
+        int64_t lastSeq = transactions.back()->getSeqNum();
+        if (tx->getSeqNum() < firstSeq || tx->getSeqNum() > lastSeq + 1)
+        {
+            return TransactionQueue::AddResult::ADD_STATUS_ERROR;
+        }
+
+        assert(tx->getSeqNum() - firstSeq <= transactions.size());
+        oldTxIter = transactions.begin() + (tx->getSeqNum() - firstSeq);
+        assert(oldTxIter == transactions.end() ||
+               (*oldTxIter)->getSeqNum() == tx->getSeqNum());
+    }
+
+    int64_t fee = tx->getFeeBid();
+    int64_t netFee = fee;
+    if (oldTxIter != transactions.end())
+    {
+        int64_t oldFee = (*oldTxIter)->getFeeBid();
+        if (fee < FEE_MULTIPLIER * oldFee)
+        {
+            return TransactionQueue::AddResult::ADD_STATUS_ERROR;
+        }
+
+        if ((*oldTxIter)->getFeeSourceID() == tx->getFeeSourceID())
+        {
+            netFee -= oldFee;
+        }
+    }
+
     LedgerTxn ltx(mApp.getLedgerTxnRoot());
-    if (!tx->checkValid(ltx, info.mMaxSeq))
     {
-        return TransactionQueue::AddResult::ADD_STATUS_ERROR;
+        auto feeSource = stellar::loadAccount(ltx, tx->getFeeSourceID());
+        if (getAvailableBalance(ltx.loadHeader(), feeSource) - netFee <
+            pending.mTotalFees)
+        {
+            tx->getResult().result.code(txINSUFFICIENT_BALANCE);
+            return TransactionQueue::AddResult::ADD_STATUS_ERROR;
+        }
     }
 
-    auto sourceAccount = stellar::loadAccount(ltx, tx->getSourceID());
-    if (getAvailableBalance(ltx.loadHeader(), sourceAccount) - tx->getFeeBid() <
-        info.mTotalFees)
+    int64_t seqNum = transactions.empty() ? 0 : (tx->getSeqNum() - 1);
+    if (!tx->checkValid(ltx, seqNum))
     {
-        tx->getResult().result.code(txINSUFFICIENT_BALANCE);
         return TransactionQueue::AddResult::ADD_STATUS_ERROR;
     }
 
-    auto& pendingForAccount = mPendingTransactions[tx->getSourceID()];
-    mSizeByAge[pendingForAccount.mAge]->inc();
-    pendingForAccount.mTotalFees += tx->getFeeBid();
-    pendingForAccount.mTransactions.emplace_back(tx);
     return TransactionQueue::AddResult::ADD_STATUS_PENDING;
 }
 
+void
+TransactionQueue::releaseFee(TransactionFrameBasePtr tx)
+{
+    auto iter = mPendingTransactions.find(tx->getFeeSourceID());
+    assert(iter != mPendingTransactions.end() &&
+           iter->second.mTotalFees >= tx->getFeeBid());
+
+    iter->second.mTotalFees -= tx->getFeeBid();
+    if (iter->second.mTransactions.empty())
```

Comment:
This seems like a footgun: `releaseFee` should probably just do that (release fee) and not manage `mPendingTransaction` or `mAge`

it looks to me that the only place where you want to clean up is in `extract` and I think the code is already there (and I don't see why you need to reset `mAge` in that case)

Reply:
cool, yes now that you've forced scopes, it seems less error prone. We probably still want to rename it